pkg/querier/tripperware/query_attribute_matcher.go

```diff
@@ -0,0 +1,218 @@
+
+	if isExpressionQuery {
+		query := r.FormValue("query")
+		expr, err := parser.ParseExpr(query)
+		if err != nil {
+			return httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+		}
+		minTime, maxTime := util.FindMinMaxTime(r, expr, lookbackDelta, now)
+
+		if queryReject := limits.QueryRejection(userStr); queryReject.Enabled && query != "" {
+			for _, attribute := range queryReject.QueryAttributes {
+				if matchAttributeForExpressionQuery(attribute, r, query, expr, now, minTime, maxTime) {
+					rejectedQueriesPerTenant.WithLabelValues(userStr).Inc()
+					return httpgrpc.Errorf(http.StatusUnprocessableEntity, queryRejectErrorMessage)
+				}
+			}
+		}
+
+		reqStats := stats.FromContext(r.Context())
+		reqStats.SetDataSelectMaxTime(maxTime)
+		reqStats.SetDataSelectMinTime(minTime)
+
+		if queryPriority := limits.QueryPriority(userStr); queryPriority.Enabled && len(queryPriority.Priorities) != 0 && query != "" {
+			for _, priority := range queryPriority.Priorities {
+				for _, attribute := range priority.QueryAttributes {
+					if matchAttributeForExpressionQuery(attribute, r, query, expr, now, minTime, maxTime) {
+						reqStats.SetPriority(priority.Priority)
+						return nil
+					}
+				}
+			}
+			reqStats.SetPriority(queryPriority.DefaultPriority)
+		}
+	}
+
+	if queryReject := limits.QueryRejection(userStr); queryReject.Enabled && isMetadataQuery {
+		for _, attribute := range queryReject.QueryAttributes {
+
+			if matchAttributeForMetadataQuery(attribute, r, now) {
+				rejectedQueriesPerTenant.WithLabelValues(userStr).Inc()
+				return httpgrpc.Errorf(http.StatusUnprocessableEntity, queryRejectErrorMessage)
+			}
+		}
+	}
+
+	return nil
+}
+
+func matchAttributeForExpressionQuery(attribute validation.QueryAttribute, r *http.Request, query string, expr parser.Expr, now time.Time, minTime, maxTime int64) bool {
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.Regex != ".+" {
+		if attribute.CompiledRegex != nil && !attribute.CompiledRegex.MatchString(query) {
+			return false
+		}
+	}
+
+	if !isWithinTimeAttributes(attribute.TimeWindow, now, minTime, maxTime) {
+		return false
+	}
+
+	if !isWithinTimeRangeAttribute(attribute.TimeRangeLimit, maxTime-minTime) {
+		return false
+	}
+
+	if !isWithinQueryStepLimit(attribute.QueryStepLimit, r, expr) {
+		return false
+	}
+
+	if attribute.UserAgent != "" && attribute.UserAgent != r.Header.Get("User-Agent") {
+		return false
+	}
+
+	if attribute.DashboardUID != "" && attribute.DashboardUID != r.Header.Get("X-Dashboard-Uid") {
+		return false
+	}
+
+	if attribute.PanelID != "" && attribute.PanelID != r.Header.Get("X-Panel-Id") {
+		return false
+	}
+
+	return true
+}
+
+func matchAttributeForMetadataQuery(attribute validation.QueryAttribute, r *http.Request, now time.Time) bool {
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.Regex != ".+" && attribute.CompiledRegex != nil {
+		for _, matcher := range r.Form["match[]"] {
```

Comment:
How can we specify API types? Like having a matcher might be a problem in `/api/v1/series` but fine for `/api/v1/labels`.

Also I don't see it distinguishes label names and label values API. They are highly likely not specifying matchers.

Reply:
I considered adding regex for URLs as well in `QueryAttribute` which could have been used to distinguish API types, but decided not to, to avoid overcomplicating it. Theoretically, it could help specify API types and narrow down matched queries, but I couldn't find a real use case. Usually we wan't to reject the queries with promql expression of specific type, independent of the queried API. 

When a regex is specified in `QueryAttribute`:
- For `query`/`query_range`, the **query** should match the regex.
- For `series`/`label`/`label_values`, at least one **match** should match the regex.

If a query to `label`/`label_values` has no **match** specified, it won't meet the `QueryAttribute` requirements and won't be rejected/prioritized. Queries with **match** or **query** matching the regex will be rejected/prioritized.

If we need rejection/prioritization based on other criteria, we can add another `QueryAttribute` without a regex but with the other fields.

Do you think distinguishing API types is necessary? Should we add a URL regex matcher?

Reply:
> If a query to label/label_values has no match specified, it won't meet the QueryAttribute requirements and won't be rejected/prioritized

I am thinking if we want to differentiate different operations. For example, requests have the same matcher `{cluster="us-west-2"}`. But we only want to reject Series API with this matcher and it is fine to allow labels and label values API using the same matcher as they are less expensive.

Reply:
added api_type property as well.